File: app/src/main/java/com/tunjid/fingergestures/gestureconsumers/BrightnessGestureConsumer.java
package com.tunjid.fingergestures.gestureconsumers;

import android.annotation.SuppressLint;
import android.content.ContentResolver;
import android.content.Context;
import android.content.ContextWrapper;
import android.content.Intent;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.provider.Settings;

import com.tunjid.fingergestures.App;
import com.tunjid.fingergestures.BrightnessLookup;
import com.tunjid.fingergestures.R;
import com.tunjid.fingergestures.SetManager;
import com.tunjid.fingergestures.activities.BrightnessActivity;
import com.tunjid.fingergestures.billing.PurchasesVerifier;

import java.math.BigDecimal;
import java.util.Comparator;
import java.util.List;
import java.util.function.Function;

import androidx.annotation.FloatRange;
import androidx.annotation.IntRange;

import static android.provider.Settings.System.SCREEN_BRIGHTNESS;
import static android.provider.Settings.System.SCREEN_BRIGHTNESS_MODE;
import static android.provider.Settings.System.SCREEN_BRIGHTNESS_MODE_AUTOMATIC;
import static android.provider.Settings.System.SCREEN_BRIGHTNESS_MODE_MANUAL;
import static com.tunjid.fingergestures.App.transformApp;
import static com.tunjid.fingergestures.App.withApp;
import static com.tunjid.fingergestures.gestureconsumers.GestureConsumer.normalizePercentageToFraction;
import static java.util.Comparator.naturalOrder;
import static java.util.Comparator.reverseOrder;

public class BrightnessGestureConsumer implements GestureConsumer {

    private static final int FUZZ_THRESHOLD = 15;
    private static final float MIN_BRIGHTNESS = 0F;
    private static final float MIN_DIM_PERCENT = 0F;
    private static final float MAX_DIM_PERCENT = .8F;
    private static final float MAX_BRIGHTNESS = 255F;
    private static final float DEF_DIM_PERCENT = MIN_DIM_PERCENT;
    private static final int MAX_ADAPTIVE_THRESHOLD = 1200;
    private static final int DEF_INCREMENT_VALUE = 20;
    private static final int DEF_POSITION_VALUE = 50;
    private static final int DEF_ADAPTIVE_BRIGHTNESS_THRESHOLD = 50;

    public static final String CURRENT_BRIGHTNESS_BYTE = "brightness value";
    public static final String ACTION_SCREEN_DIMMER_CHANGED = "show screen dimmer";
    private static final String INCREMENT_VALUE = "increment value";
    private static final String SLIDER_POSITION = "slider position";
    private static final String SLIDER_VISIBLE = "slider visible";
    private static final String LOGARITHMIC_SCALE = "logarithmic scale";
    private static final String ADAPTIVE_BRIGHTNESS = "adaptive brightness";
    private static final String ADAPTIVE_BRIGHTNESS_THRESHOLD = "adaptive brightness threshold";
    private static final String SCREEN_DIMMER_ENABLED = "screen dimmer enabled";
    private static final String SCREEN_DIMMER_DIM_PERCENT = "screen dimmer dim percent";
    private static final String ANIMATES_SLIDER = "animates slider";
    private static final String DISCRETE_BRIGHTNESS_SET = "discrete brightness values";
    private static final String EMPTY_STRING = "";

    private SetManager<Integer> discreteBrightnessManager;

    @SuppressLint("StaticFieldLeak")
    private static BrightnessGestureConsumer instance;

    public static BrightnessGestureConsumer getInstance() {
        if (instance == null) instance = new BrightnessGestureConsumer();
        return instance;
    }

    private BrightnessGestureConsumer() {
        discreteBrightnessManager = new SetManager<>(Integer::compareTo, ignored -> true, Integer::valueOf, String::valueOf);
    }

    @Override
    public void onGestureActionTriggered(@GestureAction int gestureAction) {
        int byteValue;
        int originalValue;

        App app = App.getInstance();
        if (app == null) return;

        try { byteValue = Settings.System.getInt(app.getContentResolver(), SCREEN_BRIGHTNESS); }
        catch (Exception e) {byteValue = (int) MAX_BRIGHTNESS;}

        originalValue = byteValue;

        if (gestureAction == INCREASE_BRIGHTNESS) byteValue = increase(byteValue);
        else if (gestureAction == REDUCE_BRIGHTNESS) byteValue = reduce(byteValue);
        else if (gestureAction == MAXIMIZE_BRIGHTNESS) byteValue = (int) MAX_BRIGHTNESS;
        else if (gestureAction == MINIMIZE_BRIGHTNESS) byteValue = (int) MIN_BRIGHTNESS;

        Intent intent = new Intent(app, BrightnessActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

        if (engagedDimmer(gestureAction, originalValue)) {
            byteValue = originalValue;
            intent.setAction(ACTION_SCREEN_DIMMER_CHANGED);
            intent.putExtra(SCREEN_DIMMER_DIM_PERCENT, getScreenDimmerDimPercent());
            app.broadcast(intent);
        }
        else if (shouldRemoveDimmerOnChange(gestureAction)) {
            removeDimmer();
        }

        saveBrightness(byteValue);

        intent.putExtra(CURRENT_BRIGHTNESS_BYTE, byteValue);

        if (shouldShowSlider()) app.startActivity(intent);
    }

    @Override
    @SuppressLint("SwitchIntDef")
    public boolean accepts(@GestureAction int gesture) {
        switch (gesture) {
            case INCREASE_BRIGHTNESS:
            case REDUCE_BRIGHTNESS:
            case MAXIMIZE_BRIGHTNESS:
            case MINIMIZE_BRIGHTNESS:
                return true;
            default:
                return false;
        }
    }

    public void saveBrightness(int byteValue) {
        if (!App.canWriteToSettings()) return;

        ContentResolver contentResolver = transformApp(ContextWrapper::getContentResolver);
        if (contentResolver == null) return;

        Settings.System.putInt(contentResolver, SCREEN_BRIGHTNESS_MODE, SCREEN_BRIGHTNESS_MODE_MANUAL);
        Settings.System.putInt(contentResolver, SCREEN_BRIGHTNESS, byteValue);
    }

    public void onScreenTurnedOn() {
        if (!App.canWriteToSettings()) return;

        final int threshold = adaptiveThresholdToLux(getAdaptiveBrightnessThreshold());
        final boolean restoresAdaptiveBrightness = restoresAdaptiveBrightnessOnDisplaySleep();
        final boolean toggleAndLeave = restoresAdaptiveBrightness && PurchasesVerifier.getInstance().isNotPremium();

        if (!restoresAdaptiveBrightness) return;

        if (threshold == 0 || toggleAndLeave) {
            toggleAdaptiveBrightness(true);
            return;
        }

        SensorManager sensorManager = transformApp(app -> (SensorManager) app.getSystemService(Context.SENSOR_SERVICE));
        if (sensorManager == null) return;

        final Sensor lightSensor = sensorManager.getDefaultSensor(Sensor.TYPE_LIGHT);
        if (lightSensor == null) return;

        sensorManager.registerListener(new SensorEventListener() {
            @Override
            public void onSensorChanged(SensorEvent event) {
                sensorManager.unregisterListener(this, lightSensor);

                float[] values = event.values;
                float lux = values != null && values.length > 0 ? values[0] : -1;
                boolean restoredBrightness = lux >= threshold;

                toggleAdaptiveBrightness(restoredBrightness);
                if (restoredBrightness) removeDimmer();
            }

            @Override
            public void onAccuracyChanged(Sensor sensor, int accuracy) {}
        }, lightSensor, SensorManager.SENSOR_DELAY_UI);
    }

    private boolean engagedDimmer(@GestureAction int gestureAction, int byteValue) {
        if (!isDimmerEnabled()) return false;
        if (byteValue == (int) MIN_BRIGHTNESS && gestureAction == REDUCE_BRIGHTNESS) {
            increaseScreenDimmer();
            return true;
        }
        else if (gestureAction == INCREASE_BRIGHTNESS && getScreenDimmerDimPercent() > MIN_DIM_PERCENT) {
            reduceScreenDimmer();
            return true;
        }
        return false;
    }

    private void reduceScreenDimmer() {
        float current = getScreenDimmerDimPercent();
        float changed = current - normalizePercentageToFraction(getIncrementPercentage());
        setDimmerPercent(Math.max(roundDown(changed), MIN_DIM_PERCENT));
    }

    private void increaseScreenDimmer() {
        float current = getScreenDimmerDimPercent();
        float changed = current + normalizePercentageToFraction(getIncrementPercentage());
        setDimmerPercent(Math.min(roundDown(changed), MAX_DIM_PERCENT));
    }

    private int reduce(int byteValue) {
        return noDiscreteBrightness()
                ? Math.max(adjustByte(byteValue, false), (int) MIN_BRIGHTNESS)
                : findDiscreteBrightnessValue(byteValue, false);
    }

    private int increase(int byteValue) {
        return noDiscreteBrightness()
                ? Math.min(adjustByte(byteValue, true), (int) MAX_BRIGHTNESS)
                : findDiscreteBrightnessValue(byteValue, true);
    }

    private int adjustByte(int byteValue, boolean increase) {
        int delta = getIncrementPercentage();
        int asPercentage = byteToPercentage(byteValue);
        if (increase) asPercentage = Math.min(asPercentage + delta, 100);
        else asPercentage = Math.max(asPercentage - delta, 0);

        return percentToByte(asPercentage);
    }

    public void setIncrementPercentage(@IntRange(from = GestureConsumer.ZERO_PERCENT,
            to = GestureConsumer.HUNDRED_PERCENT) int incrementValue) {
        withApp(app -> app.getPreferences().edit().putInt(INCREMENT_VALUE, incrementValue).apply());
    }

    public void setPositionPercentage(@IntRange(from = GestureConsumer.ZERO_PERCENT,
            to = GestureConsumer.HUNDRED_PERCENT) int positionPercentage) {
        withApp(app -> app.getPreferences().edit().putInt(SLIDER_POSITION, positionPercentage).apply());
    }

    private void setDimmerPercent(@FloatRange(from = GestureConsumer.ZERO_PERCENT,
            to = GestureConsumer.HUNDRED_PERCENT) float percentage) {
        withApp(app -> app.getPreferences().edit().putFloat(SCREEN_DIMMER_DIM_PERCENT, percentage).apply());
    }

    public void setAdaptiveBrightnessThreshold(@IntRange(from = GestureConsumer.ZERO_PERCENT,
            to = GestureConsumer.HUNDRED_PERCENT) int threshold) {
        withApp(app -> app.getPreferences().edit().putInt(ADAPTIVE_BRIGHTNESS_THRESHOLD, threshold).apply());
    }

    public void shouldRestoreAdaptiveBrightnessOnDisplaySleep(boolean restore) {
        withApp(app -> app.getPreferences().edit().putBoolean(ADAPTIVE_BRIGHTNESS, restore).apply());
    }

    public void shouldUseLogarithmicScale(boolean isLogarithmic) {
        withApp(app -> app.getPreferences().edit().putBoolean(LOGARITHMIC_SCALE, isLogarithmic).apply());
    }

    public void setSliderVisible(boolean visible) {
        withApp(app -> app.getPreferences().edit().putBoolean(SLIDER_VISIBLE, visible).apply());
    }

    public void setAnimatesSlider(boolean visible) {
        withApp(app -> app.getPreferences().edit().putBoolean(ANIMATES_SLIDER, visible).apply());
    }

    public void setDimmerEnabled(boolean enabled) {
        withApp(app -> app.getPreferences().edit().putBoolean(SCREEN_DIMMER_ENABLED, enabled).apply());
        if (!enabled) removeDimmer();
    }

    @IntRange(from = GestureConsumer.ZERO_PERCENT, to = GestureConsumer.HUNDRED_PERCENT)
    public int getIncrementPercentage() {
        return transformApp(app -> app.getPreferences().getInt(INCREMENT_VALUE, DEF_INCREMENT_VALUE), DEF_INCREMENT_VALUE);
    }

    @IntRange(from = GestureConsumer.ZERO_PERCENT, to = GestureConsumer.HUNDRED_PERCENT)
    public int getPositionPercentage() {
        return transformApp(app -> app.getPreferences().getInt(SLIDER_POSITION, DEF_POSITION_VALUE), DEF_POSITION_VALUE);
    }

    @IntRange(from = GestureConsumer.ZERO_PERCENT, to = GestureConsumer.HUNDRED_PERCENT)
    public int getAdaptiveBrightnessThreshold() {
        return transformApp(app -> app.getPreferences().getInt(ADAPTIVE_BRIGHTNESS_THRESHOLD, DEF_ADAPTIVE_BRIGHTNESS_THRESHOLD), DEF_ADAPTIVE_BRIGHTNESS_THRESHOLD);
    }

    public float getScreenDimmerDimPercent() {
        return transformApp(app -> app.getPreferences().getFloat(SCREEN_DIMMER_DIM_PERCENT, DEF_DIM_PERCENT), DEF_DIM_PERCENT);
    }

    public boolean restoresAdaptiveBrightnessOnDisplaySleep() {
        return transformApp(app -> app.getPreferences().getBoolean(ADAPTIVE_BRIGHTNESS, false), false);
    }

    public boolean usesLogarithmicScale() {
        return transformApp(app -> app.getPreferences().getBoolean(LOGARITHMIC_SCALE, App.isPieOrHigher()), App.isPieOrHigher());
    }

    public boolean supportsAmbientThreshold() {
        return PurchasesVerifier.getInstance().isPremium()
                && restoresAdaptiveBrightnessOnDisplaySleep()
                && hasBrightnessSensor();
    }

    public boolean hasOverlayPermission() {
        return transformApp(Settings::canDrawOverlays, false);
    }

    public boolean isDimmerEnabled() {
        return hasOverlayPermission()
                && PurchasesVerifier.getInstance().isPremium()
                && transformApp(app -> app.getPreferences().getBoolean(SCREEN_DIMMER_ENABLED, false), false);
    }

    public boolean shouldShowDimmer() {
        return getScreenDimmerDimPercent() != MIN_DIM_PERCENT;
    }

    public boolean shouldShowSlider() {
        return transformApp(app -> app.getPreferences().getBoolean(SLIDER_VISIBLE, true), true);
    }

    public boolean shouldAnimateSlider() {
        return transformApp(app -> app.getPreferences().getBoolean(ANIMATES_SLIDER, true), true);
    }

    public boolean canAdjustDelta() {
        return noDiscreteBrightness() || PurchasesVerifier.getInstance().isPremium();
    }

    public void addDiscreteBrightnessValue(String discreteValue) {
        discreteBrightnessManager.addToSet(discreteValue, DISCRETE_BRIGHTNESS_SET);
    }

    public void removeDiscreteBrightnessValue(String discreteValue) {
        discreteBrightnessManager.removeFromSet(discreteValue, DISCRETE_BRIGHTNESS_SET);
    }

    public String getAdjustDeltaText(int percentage) {
        return transformApp(app -> PurchasesVerifier.getInstance().isPremium() && !noDiscreteBrightness()
                ? app.getString(R.string.delta_percent_premium, percentage)
                : app.getString(R.string.delta_percent, percentage), EMPTY_STRING);
    }

    public List<String> getDiscreteBrightnessValues() {
        return discreteBrightnessManager.getList(DISCRETE_BRIGHTNESS_SET);
    }

    public String getAdaptiveBrightnessThresholdText(@IntRange(from = GestureConsumer.ZERO_PERCENT,
            to = GestureConsumer.HUNDRED_PERCENT) int percent) {
        if (!hasBrightnessSensor())
            return transformApp(app -> app.getString(R.string.unavailable_brightness_sensor), EMPTY_STRING);

        if (PurchasesVerifier.getInstance().isNotPremium())
            return transformApp(app -> app.getString(R.string.go_premium_text), EMPTY_STRING);

        if (!restoresAdaptiveBrightnessOnDisplaySleep())
            return transformApp(app -> app.getString(R.string.adjust_adaptive_threshold_prompt), EMPTY_STRING);

        int lux = adaptiveThresholdToLux(percent);

        return transformApp(app -> {
            String descriptor = app.getString(lux < 50
                    ? R.string.adaptive_threshold_low
                    : lux < 1000 ? R.string.adaptive_threshold_medium
                    : R.string.adaptive_threshold_high);

            return app.getString(R.string.adaptive_threshold, lux, descriptor);
        }, EMPTY_STRING);
    }

    public void removeDimmer() {
        setDimmerPercent(MIN_DIM_PERCENT);
        Intent intent = new Intent(ACTION_SCREEN_DIMMER_CHANGED);
        intent.putExtra(SCREEN_DIMMER_DIM_PERCENT, getScreenDimmerDimPercent());
        withApp(app -> app.broadcast(intent));
    }

    public int percentToByte(int percentage) {
        return usesLogarithmicScale()
                ? BrightnessLookup.lookup(percentage, false)
                : (int) (percentage * MAX_BRIGHTNESS / 100);
    }

    public int byteToPercentage(int byteValue) {
        return usesLogarithmicScale()
                ? BrightnessLookup.lookup(byteValue, true)
                : (int) (byteValue * 100 / MAX_BRIGHTNESS);
    }

    private int adaptiveThresholdToLux(int percentage) {
        return (int) (percentage * MAX_ADAPTIVE_THRESHOLD / 100F);
    }

    private int findDiscreteBrightnessValue(int byteValue, boolean increasing) {
        int evaluated = byteValue < FUZZ_THRESHOLD ? byteValue : increasing ? byteValue + 2 : byteValue - 2;
        int alternative = (int) (increasing ? MAX_BRIGHTNESS : MIN_BRIGHTNESS);
        Comparator<Integer> comparator = increasing ? naturalOrder() : reverseOrder();
        Function<Integer, Boolean> filter = integer -> increasing ? integer > evaluated : integer < evaluated;

        return discreteBrightnessManager.getSet(DISCRETE_BRIGHTNESS_SET).stream()
                .map(this::stringPercentageToBrightnessInt)
                .filter(filter::apply).min(comparator).orElse(alternative);
    }

    private void toggleAdaptiveBrightness(boolean on) {
        int brightnessMode = on
                ? SCREEN_BRIGHTNESS_MODE_AUTOMATIC
                : SCREEN_BRIGHTNESS_MODE_MANUAL;
        withApp(app -> Settings.System.putInt(app.getContentResolver(), SCREEN_BRIGHTNESS_MODE, brightnessMode));
    }

    private static float roundDown(float d) {
        BigDecimal bd = new BigDecimal(Float.toString(d));
        bd = bd.setScale(2, BigDecimal.ROUND_HALF_DOWN);
        return bd.floatValue();
    }

    private int stringPercentageToBrightnessInt(String stringPercent) {
        Float percentage = Float.valueOf(stringPercent);
        return percentToByte(percentage.intValue());
    }

    private boolean hasBrightnessSensor() {
        SensorManager sensorManager = transformApp(app -> (SensorManager) app.getSystemService(Context.SENSOR_SERVICE));
        if (sensorManager == null) return false;

        final Sensor lightSensor = sensorManager.getDefaultSensor(Sensor.TYPE_LIGHT);
        return lightSensor != null;
    }

    private boolean shouldRemoveDimmerOnChange(@GestureAction int gestureAction) {
        return gestureAction == MINIMIZE_BRIGHTNESS || gestureAction == MAXIMIZE_BRIGHTNESS
                || (shouldShowDimmer() && PurchasesVerifier.getInstance().isNotPremium());
    }

    private boolean noDiscreteBrightness() {
        return discreteBrightnessManager.getSet(DISCRETE_BRIGHTNESS_SET).isEmpty();
    }
}

File: app/src/main/java/com/tunjid/fingergestures/gestureconsumers/BrightnessGestureConsumer.kt
/*
 * Copyright (c) 2017, 2018, 2019 Adetunji Dahunsi.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.tunjid.fingergestures.gestureconsumers

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.provider.Settings
import androidx.annotation.IntRange
import com.jakewharton.rx.replayingShare
import com.tunjid.androidx.core.delegates.intentExtras
import com.tunjid.fingergestures.*
import com.tunjid.fingergestures.activities.BrightnessActivity
import com.tunjid.fingergestures.billing.PurchasesManager
import com.tunjid.fingergestures.di.AppBroadcaster
import com.tunjid.fingergestures.di.AppContext
import com.tunjid.fingergestures.di.AppDisposable
import com.tunjid.fingergestures.models.Broadcast
import io.reactivex.Flowable
import io.reactivex.rxkotlin.Flowables
import java.math.BigDecimal
import java.util.*
import java.util.Comparator.naturalOrder
import java.util.Comparator.reverseOrder
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.max
import kotlin.math.min

@Singleton
class BrightnessGestureConsumer @Inject constructor(
    @AppContext private val context: Context,
    reactivePreferences: ReactivePreferences,
    private val broadcaster: AppBroadcaster,
    private val purchasesManager: PurchasesManager,
    appDisposable: AppDisposable
) : GestureConsumer {

    data class SliderPair(
        val value: Int,
        val enabled: Boolean,
    )

    data class DimmerState(
        val enabled: Boolean = false,
        val visible: Boolean = false,
        val checked: Boolean = false,
        val percentage: Float = 0f,
    )

    data class State(
        val increment: SliderPair = SliderPair(value = DEF_INCREMENT_VALUE, enabled = true),
        val position: SliderPair = SliderPair(value = DEF_POSITION_VALUE, enabled = true),
        val adaptive: SliderPair = SliderPair(value = DEF_ADAPTIVE_BRIGHTNESS_THRESHOLD, enabled = true),
        val dimmerState: DimmerState = DimmerState(),
        val restoresAdaptiveBrightnessOnDisplaySleep: Boolean = true,
        val usesLogarithmicScale: Boolean = App.isPieOrHigher,
        val shouldShowSlider: Boolean = true,
        val shouldAnimateSlider: Boolean = true,
        val discreteBrightnesses: List<Int> = listOf(),
    )

    enum class Preference(override val preferenceName: String) : SetPreference {
        DiscreteBrightnesses(preferenceName = "discrete brightness values");
    }

    val percentagePreference: ReactivePreference<Int> = ReactivePreference(
        reactivePreferences = reactivePreferences,
        key = "increment value",
        default = DEF_INCREMENT_VALUE
    )
    val positionPreference: ReactivePreference<Int> = ReactivePreference(
        reactivePreferences = reactivePreferences,
        key = "slider position",
        default = DEF_POSITION_VALUE
    )
    val adaptiveBrightnessThresholdPreference: ReactivePreference<Int> = ReactivePreference(
        reactivePreferences = reactivePreferences,
        key = "adaptive brightness threshold",
        default = DEF_ADAPTIVE_BRIGHTNESS_THRESHOLD
    )
    val adaptiveBrightnessPreference: ReactivePreference<Boolean> = ReactivePreference(
        reactivePreferences = reactivePreferences,
        key = "adaptive brightness",
        default = true
    )
    val logarithmicBrightnessPreference: ReactivePreference<Boolean> = ReactivePreference(
        reactivePreferences = reactivePreferences,
        key = "logarithmic scale",
        default = App.isPieOrHigher
    )
    val showSliderPreference: ReactivePreference<Boolean> = ReactivePreference(
        reactivePreferences = reactivePreferences,
        key = "slider visible",
        default = true
    )
    val animateSliderPreference: ReactivePreference<Boolean> = ReactivePreference(
        reactivePreferences = reactivePreferences,
        key = "animates slider",
        default = true
    )
    val screenDimmerPercentPreference: ReactivePreference<Float> = ReactivePreference(
        reactivePreferences = reactivePreferences,
        key = "screen dimmer dim percent",
        default = DEF_DIM_PERCENT
    )
    val screenDimmerEnabledPreference: ReactivePreference<Boolean> = ReactivePreference(
        reactivePreferences = reactivePreferences,
        key = "screen dimmer enabled",
        default = false,
        onSet = { enabled -> if (!enabled) removeDimmer() }
    )
    val discreteBrightnessManager: SetManager<Preference, Int> = SetManager(
        reactivePreferences = reactivePreferences,
        keys = Preference.values().toList(),
        sorter = Comparator(Int::compareTo),
        addFilter = { true },
        stringMapper = Integer::valueOf,
        objectMapper = Int::toString
    )

    private val isDimmerEnabled: Boolean
        get() = hasOverlayPermission
            && purchasesManager.currentState.isPremium
            && currentState.dimmerState.checked

    private val hasOverlayPermission: Boolean
        get() = Settings.canDrawOverlays(context)

    private val supportsAmbientThreshold = Flowables.combineLatest(
        purchasesManager.state,
        adaptiveBrightnessPreference.monitor
    ) { purchaseState, restores -> purchaseState.isPremium && restores && hasBrightnessSensor() }

    val state: Flowable<State> = Flowables.combineLatest(
        Flowables.combineLatest(
            discreteBrightnessManager.itemsFlowable(Preference.DiscreteBrightnesses),
            purchasesManager.state,
            percentagePreference.monitor,

            ) { discreteBrightnesses, purchaseState, percentage ->
            SliderPair(value = percentage, enabled = purchaseState.isPremium || discreteBrightnesses.isEmpty())
        },
        Flowables.combineLatest(
            positionPreference.monitor,
            Flowable.just(true),
            ::SliderPair,
        ),
        Flowables.combineLatest(
            adaptiveBrightnessThresholdPreference.monitor,
            supportsAmbientThreshold,
            ::SliderPair,
        ),
        Flowables.combineLatest(
            purchasesManager.state,
            screenDimmerEnabledPreference.monitor,
            screenDimmerPercentPreference.monitor
        ) { purchaseState, dimmerEnabled, dimmerPercent ->
            DimmerState(
                enabled = purchaseState.isPremium,
                // TODO: Make reactive
                visible = hasOverlayPermission,
                checked = dimmerEnabled,
                percentage = dimmerPercent
            )
        },
        adaptiveBrightnessPreference.monitor,
        logarithmicBrightnessPreference.monitor,
        showSliderPreference.monitor,
        animateSliderPreference.monitor,
        discreteBrightnessManager.itemsFlowable(Preference.DiscreteBrightnesses),
        ::State,
    ).replayingShare()

    private val currentState by state.asProperty(State(), appDisposable::add)

    override fun onGestureActionTriggered(gestureAction: GestureAction) {
        var byteValue: Int
        val originalValue: Int

        byteValue = try {
            Settings.System.getInt(context.contentResolver, Settings.System.SCREEN_BRIGHTNESS)
        } catch (e: Exception) {
            MAX_BRIGHTNESS.toInt()
        }

        originalValue = byteValue

        when (gestureAction) {
            GestureAction.IncreaseBrightness -> byteValue = increase(byteValue)
            GestureAction.ReduceBrightness -> byteValue = reduce(byteValue)
            GestureAction.MaximizeBrightness -> byteValue = MAX_BRIGHTNESS.toInt()
            GestureAction.MinimizeBrightness -> byteValue = MIN_BRIGHTNESS.toInt()
            else -> Unit
        }

        val intent = Intent(context, BrightnessActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK

        if (engagedDimmer(gestureAction, originalValue)) {
            byteValue = originalValue
            // Read it directly, can't optimize by checking current
            val percent = screenDimmerPercentPreference.value
            intent.action = ACTION_SCREEN_DIMMER_CHANGED
            intent.screenDimmerPercent = percent
            broadcaster(Broadcast.Service.ScreenDimmerChanged(percent))
        } else if (shouldRemoveDimmerOnChange(gestureAction)) removeDimmer()

        saveBrightness(byteValue)

        intent.putExtra(CURRENT_BRIGHTNESS_BYTE, byteValue)

        if (showSliderPreference.value) context.startActivity(intent)
    }

    @SuppressLint("SwitchIntDef")
    override fun accepts(gesture: GestureAction): Boolean = when (gesture) {
        GestureAction.IncreaseBrightness,
        GestureAction.ReduceBrightness,
        GestureAction.MaximizeBrightness,
        GestureAction.MinimizeBrightness -> true
        else -> false
    }

    val percentageFormatter = { percent: Int -> context.getString(R.string.position_percent, percent) }

    fun saveBrightness(byteValue: Int) {
        if (!context.canWriteToSettings) return

        val contentResolver = context.contentResolver ?: return

        Settings.System.putInt(contentResolver, Settings.System.SCREEN_BRIGHTNESS_MODE, Settings.System.SCREEN_BRIGHTNESS_MODE_MANUAL)
        Settings.System.putInt(contentResolver, Settings.System.SCREEN_BRIGHTNESS, byteValue)
    }

    fun onScreenTurnedOn() {
        if (!context.canWriteToSettings) return

        val threshold = adaptiveThresholdToLux(adaptiveBrightnessThresholdPreference.value)
        val restoresAdaptiveBrightness = adaptiveBrightnessPreference.value
        val toggleAndLeave = restoresAdaptiveBrightness && !purchasesManager.currentState.isPremium

        if (!restoresAdaptiveBrightness) return

        if (threshold == 0 || toggleAndLeave) {
            toggleAdaptiveBrightness(true)
            return
        }

        val sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as? SensorManager
            ?: return

        val lightSensor = sensorManager.getDefaultSensor(Sensor.TYPE_LIGHT) ?: return

        sensorManager.registerListener(object : SensorEventListener {
            override fun onSensorChanged(event: SensorEvent) {
                sensorManager.unregisterListener(this, lightSensor)

                val values = event.values
                val lux = if (values != null && values.isNotEmpty()) values[0] else -1F
                val restoredBrightness = lux >= threshold

                toggleAdaptiveBrightness(restoredBrightness)
                if (restoredBrightness) removeDimmer()
            }

            override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}
        }, lightSensor, SensorManager.SENSOR_DELAY_UI)
    }

    private fun engagedDimmer(gestureAction: GestureAction, byteValue: Int): Boolean =
        when {
            !isDimmerEnabled -> false
            byteValue == MIN_BRIGHTNESS.toInt() && gestureAction == GestureAction.ReduceBrightness -> {
                increaseScreenDimmer()
                true
            }
            gestureAction == GestureAction.IncreaseBrightness && currentState.dimmerState.percentage > MIN_DIM_PERCENT -> {
                reduceScreenDimmer()
                true
            }
            else -> false
        }

    private fun reduceScreenDimmer() {
        val current = currentState.dimmerState.percentage
        val changed = current - GestureConsumer.normalizePercentageToFraction(percentagePreference.value)
        screenDimmerPercentPreference.value = max(roundDown(changed), MIN_DIM_PERCENT)
    }

    private fun increaseScreenDimmer() {
        val current = currentState.dimmerState.percentage
        val changed = current + GestureConsumer.normalizePercentageToFraction(percentagePreference.value)
        screenDimmerPercentPreference.value = min(roundDown(changed), MAX_DIM_PERCENT)
    }

    private fun reduce(byteValue: Int): Int =
        if (noDiscreteBrightness()) max(adjustByte(byteValue, false), MIN_BRIGHTNESS.toInt())
        else findDiscreteBrightnessValue(byteValue, false)

    private fun increase(byteValue: Int): Int =
        if (noDiscreteBrightness()) min(adjustByte(byteValue, true), MAX_BRIGHTNESS.toInt())
        else findDiscreteBrightnessValue(byteValue, true)

    private fun adjustByte(byteValue: Int, increase: Boolean): Int {
        val delta = percentagePreference.value
        var asPercentage = byteToPercentage(byteValue)

        asPercentage = if (increase) min(asPercentage + delta, 100)
        else max(asPercentage - delta, 0)

        return percentToByte(asPercentage)
    }

    fun shouldShowDimmer(): Boolean = currentState.dimmerState.percentage != MIN_DIM_PERCENT

    fun getAdjustDeltaText(percentage: Int): String = context.let { app ->
        if (purchasesManager.currentState.isPremium && !noDiscreteBrightness())
            app.getString(R.string.delta_percent_premium, percentage)
        else app.getString(R.string.delta_percent, percentage)
    }

    fun getAdaptiveBrightnessThresholdText(@IntRange(from = GestureConsumer.ZERO_PERCENT.toLong(), to = GestureConsumer.HUNDRED_PERCENT.toLong())
                                           percent: Int): String {
        if (!hasBrightnessSensor())
            return context.getString(R.string.unavailable_brightness_sensor)

        if (!purchasesManager.currentState.isPremium)
            return context.getString(R.string.go_premium_text)

        if (!adaptiveBrightnessPreference.value)
            return context.getString(R.string.adjust_adaptive_threshold_prompt)

        val lux = adaptiveThresholdToLux(percent)
        val descriptor = context.getString(when {
            lux < 50 -> R.string.adaptive_threshold_low
            lux < 1000 -> R.string.adaptive_threshold_medium
            else -> R.string.adaptive_threshold_high
        })

        return context.getString(R.string.adaptive_threshold, lux, descriptor)
    }

    fun removeDimmer() {
        screenDimmerPercentPreference.value = MIN_DIM_PERCENT
        broadcaster(Broadcast.Service.ScreenDimmerChanged(MIN_DIM_PERCENT))
    }

    fun percentToByte(percentage: Int): Int =
        if (logarithmicBrightnessPreference.value) BrightnessLookup.lookup(percentage, false)
        else (percentage * MAX_BRIGHTNESS / 100).toInt()

    fun byteToPercentage(byteValue: Int): Int =
        if (logarithmicBrightnessPreference.value) BrightnessLookup.lookup(byteValue, true)
        else (byteValue * 100 / MAX_BRIGHTNESS).toInt()

    private fun adaptiveThresholdToLux(percentage: Int): Int =
        (percentage * MAX_ADAPTIVE_THRESHOLD / 100f).toInt()

    private fun findDiscreteBrightnessValue(byteValue: Int, increasing: Boolean): Int {
        val evaluated = if (byteValue < FUZZ_THRESHOLD) byteValue else if (increasing) byteValue + 2 else byteValue - 2
        val alternative = (if (increasing) MAX_BRIGHTNESS else MIN_BRIGHTNESS).toInt()
        val comparator = if (increasing) naturalOrder() else reverseOrder<Int>()
        val filter = { integer: Int -> if (increasing) integer > evaluated else integer < evaluated }

        return discreteBrightnessManager.getSet(Preference.DiscreteBrightnesses).stream()
            .map { this.stringPercentageToBrightnessInt(it) }
            .filter(filter::invoke).min(comparator).orElse(alternative)
    }

    private fun toggleAdaptiveBrightness(on: Boolean) {
        val brightnessMode =
            if (on) Settings.System.SCREEN_BRIGHTNESS_MODE_AUTOMATIC
            else Settings.System.SCREEN_BRIGHTNESS_MODE_MANUAL

        Settings.System.putInt(context.contentResolver, Settings.System.SCREEN_BRIGHTNESS_MODE, brightnessMode)
    }

    private fun stringPercentageToBrightnessInt(stringPercent: String): Int {
        val percentage = java.lang.Float.valueOf(stringPercent)
        return percentToByte(percentage.toInt())
    }

    private fun hasBrightnessSensor(): Boolean {
        val sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as? SensorManager
            ?: return false

        val lightSensor = sensorManager.getDefaultSensor(Sensor.TYPE_LIGHT)
        return lightSensor != null
    }

    private fun shouldRemoveDimmerOnChange(gestureAction: GestureAction): Boolean =
        (gestureAction == GestureAction.MinimizeBrightness || gestureAction == GestureAction.MaximizeBrightness
            || shouldShowDimmer() && !purchasesManager.currentState.isPremium)

    private fun noDiscreteBrightness(): Boolean =
        discreteBrightnessManager.getSet(Preference.DiscreteBrightnesses).isEmpty()

    companion object {

        private const val FUZZ_THRESHOLD = 15
        private const val MIN_BRIGHTNESS = 0f
        private const val MIN_DIM_PERCENT = 0f
        private const val MAX_DIM_PERCENT = .8f
        private const val MAX_BRIGHTNESS = 255f
        private const val DEF_DIM_PERCENT = MIN_DIM_PERCENT
        private const val MAX_ADAPTIVE_THRESHOLD = 1200
        private const val DEF_INCREMENT_VALUE = 20
        private const val DEF_POSITION_VALUE = 50
        private const val DEF_ADAPTIVE_BRIGHTNESS_THRESHOLD = 50

        const val CURRENT_BRIGHTNESS_BYTE = "brightness value"
        const val ACTION_SCREEN_DIMMER_CHANGED = "show screen dimmer"

        private fun roundDown(d: Float): Float {
            var bd = BigDecimal(d.toString())
            bd = bd.setScale(2, BigDecimal.ROUND_HALF_DOWN)
            return bd.toFloat()
        }
    }
}

var Intent.screenDimmerPercent by intentExtras<Float?>()